monai/data/wsi_reader.py

```diff
@@ -0,0 +1,428 @@
+
+    @abstractmethod
+    def get_patch(self, wsi, location: Tuple[int, int], size: Tuple[int, int], level: int, dtype: DtypeLike, mode: str):
+        """
+        Extracts and returns a patch image form the whole slide image.
+
+        Args:
+            wsi: a whole slide image object loaded from a file or a lis of such objects
+            location: (x_min, y_min) tuple giving the top left pixel in the level 0 reference frame. Defaults to (0, 0).
+            size: (height, width) tuple giving the patch size at the given level (`level`).
+                If None, it is set to the full image size at the given level.
+            level: the level number. Defaults to 0
+            dtype: the data type of output image
+            mode: the output image mode, 'RGB' or 'RGBA'
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+    @abstractmethod
+    def get_metadata(self, patch: np.ndarray, location: Tuple[int, int], size: Tuple[int, int], level: int):
+        """
+        Extracts and returns metadata form the whole slide image.
+
+        Args:
+            patch: extracted patch from whole slide image
+            location: (x_min, y_min) tuple giving the top left pixel in the level 0 reference frame. Defaults to (0, 0).
+            size: (height, width) tuple giving the patch size at the given level (`level`).
+                If None, it is set to the full image size at the given level.
+            level: the level number. Defaults to 0
+
+        """
+        raise NotImplementedError(f"Subclass {self.__class__.__name__} must implement this method.")
+
+    def get_data(
+        self,
+        wsi,
+        location: Tuple[int, int] = (0, 0),
+        size: Optional[Tuple[int, int]] = None,
+        level: Optional[int] = None,
+        dtype: DtypeLike = np.uint8,
+        mode: str = "RGB",
+    ):
+        """
+        Verifies inputs, extracts patches from WSI image and generates metadata, and return them.
+
+        Args:
+            wsi: a whole slide image object loaded from a file or a lis of such objects
+            location: (x_min, y_min) tuple giving the top left pixel in the level 0 reference frame. Defaults to (0, 0).
+            size: (height, width) tuple giving the patch size at the given level (`level`).
+                If None, it is set to the full image size at the given level.
+            level: the level number. Defaults to 0
+            dtype: the data type of output image
+            mode: the output image mode, 'RGB' or 'RGBA'
+
+        Returns:
+            a tuples, where the first element is an image patch [CxHxW] or stack of patches,
+                and second element is a dictionary of metadata
+        """
+        patch_list: List = []
+        metadata = {}
+        # CuImage object is iterable, so ensure_tuple won't work on single object
+        if not isinstance(wsi, List):
```

Comment:
I think no need these 2 lines, `ensure_tuple()` can automatically support a single object, just like your code:
https://github.com/Project-MONAI/MONAI/pull/4107/files#diff-e0abb01781c5bed5c90cf46759d8e2aeac0793b86bd48ab49b40b6b7f7f5c1c2R382

Thanks.

Reply:
These lines are essential since `ensure_tuple` gets confused when the object itself is iterable.

Reply:
Oh, I see. I didn't know the `wsi` is an iterable object.

Thanks for the explanation.